File: nvpl_blas/f77/ssyr2k.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     SSYR2K 
!
!******************************************************************************
      program   SSYR2K_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1     trans, uplo
      parameter       (trans='N', uplo='U')
      integer         n, k, i, j
      integer         lda, ldb, ldc
      parameter       (n=2, k=2,
     &                 lda=2, ldb=2, ldc=2)
      integer         asize, bsize, csize
      parameter       (asize=n*k, bsize=n*k, csize=n*n)
      real            alpha, beta
      parameter       (alpha=1.0, beta=1.0)
      real            a(lda,k), b(ldb,n), c(ldc,n)

*       External Subroutines
      external        print_smatrix, fill_smatrix
*
*      Executable Statements
*
      print*
      print 99, 'Example: SSYR2K for the symmetric rank-2k update'
*
*       Print input data
      print*
      print 100, n, k, lda, ldb, ldc, trans, uplo
      print 101, alpha, beta
      call fill_smatrix(a,n,k,lda,'F','N')
      call fill_smatrix(b,n,k,ldb,'F','N')
      call fill_smatrix(c,n,n,ldc,uplo,'N')
      call print_smatrix(a,n,k,lda,'A')
      call print_smatrix(b,n,k,ldb,'B')
      call print_smatrix(c,n,n,ldc,'C')
*
*      Call SSYR2K subroutine
      call SSYR2K(uplo,trans,n,k,alpha,a,lda,b,ldb,beta,c,ldc)
*
      print*
      print 102, 'After SSYR2K operation:'
      call print_smatrix(c,n,n,ldc,'C')

      stop
 99   format(A)
 100  format('#### args: n=',i1,', k=',i1,', lda=',i1,
     &       ', ldb=',i1,', ldc=',i1,', trans=',a1,', uplo=',a1)
 101  format(11x,'alpha=',f4.1,'  beta=',f4.1)
 102  format(A)
      end

File: nvpl_blas/c/ssyr2k.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_ssyr2k
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t K = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    nvpl_int_t ldc = 2;
    float alpha = 1.0f;
    float beta = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE trans = CblasNoTrans;
    enum CBLAS_UPLO uplo = CblasUpper;
    float * A = NULL;
    float * B = NULL;
    float * C = NULL;

    printf("\nExample: cblas_ssyr2k for the symmetric rank-2k update\n\n");
    printf("#### args: n=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", ldc=%" PRId64 ", "
            "trans=%c, uplo=%c, order=%c\n", (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)ldb,
            (int64_t)ldc, transpose_to_char(trans), uplo_to_char(uplo), order_to_char(order));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    nvpl_int_t rowsA = (trans == CblasNoTrans) ? N : K;
    nvpl_int_t colsA = (trans == CblasNoTrans) ? K : N;
    nvpl_int_t rowsB = (trans == CblasNoTrans) ? N : K;
    nvpl_int_t colsB = (trans == CblasNoTrans) ? K : N;
    nvpl_int_t rowsC = N;
    nvpl_int_t colsC = N;

    // allocate memory
    A = (float *)malloc(lda * colsA  * sizeof(float));
    B = (float *)malloc(ldb * colsB  * sizeof(float));
    C = (float *)malloc(ldc * colsC  * sizeof(float));

    // fill data
    fill_smatrix(A, rowsA, colsA, lda, order, Full, CblasNonUnit);
    fill_smatrix(B, rowsB, colsB, ldb, order, Full, CblasNonUnit);
    fill_smatrix(C, rowsC, colsC, ldc, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);

    // print input data
    print_smatrix(A, rowsA, colsA, lda, "A", order);
    print_smatrix(B, rowsB, colsB, ldb, "B", order);
    print_smatrix(C, rowsC, colsC, ldc, "C", order);

    // call cblas_ssyr2k
    cblas_ssyr2k(order, uplo, trans, N, K, alpha, A, lda, B, ldb, beta, C, ldc);

    // print result
    printf("\nAfter ssyr2k operation:\n");
    print_smatrix(C, rowsC, colsC, ldc, "C", order);

    // release memory
    free(A);
    free(B);
    free(C);
    return 0;
}
